### ISSUE TITLE
Empty objects break Dynamo

### ISSUE THREAD

**Issue by numberoverzero** [numberoverzero/bloop#49]
# tl;dr

Unfortunately, this spans the type system, tracking, and the condition renderer.

I'm going to try and hold off releasing this until 1.0.0 because it's a significant change internally, and it's a large change to the user-facing contract of `bloop.Type`.  In the meantime, you'll need to set vector types (List, Set, Map, TypedMap) to `None` before saving if they are empty.  That is, if you have the following object:

```python
instance = Model(id=0, data=set())
engine.save(instance)
```

it needs to do the following:

```python
instance = Model(id=0, data=set())

# check for empty sets before saving!
# If the set is empty, set it to None to
# avoid validation errors on passing an empty set
instance.data = instance.data or None

engine.save(instance)
```
## TODO

- [x] drop `tracking.get_update` for `tracking.marked_columns`
- [x] `engine.save` calls `renderer.update(obj)` instead of `diff = tracking.get_update(obj)`
- [x] `ConditionRenderer.update` takes `Set[Column]` and builds values into correct `SET` or `REMOVE` list
- [x] all types can handle `dynamo_load(None, ...)`
- [x] `types.Type._load` doesn't guard against `None`
- [x] all types can handle `dynamo_dump(None, ...)`
- [x] vector types (List/Set/Map/TypedMap) handle `None` when dumping inner types
- [x] `types.Type._dump` doesn't guard against `None`
- [x] piles of documentation need to be rewritten

## Repro

```python
from bloop import new_base, Column, Integer, Set, Engine


class Model(new_base()):
    id = Column(Integer, hash_key=True)
    data = Column(Set(Integer))

engine = Engine()
engine.bind(base=Model)

instance = Model(id=2)


# This save is fine; data isn't set
engine.save(instance)
print("First call succeeded")


# This save fails; data is an empty set()
instance.data = set()
engine.save(instance)
```

#### `types.Set`

This is because the type will return an empty list when dumped:

```python

typedef = Set(Integer)
context = {"engine": engine}
assert typedef.dump(set(), context=context) == []
```

Which is rendered (eventually, in some `ExpressionAttributeValues`) as `{"SS": []}` and the UpdateItem fails with an error message explaining that we can't have an empty set for a value:

```
botocore.exceptions.ClientError: An error occurred (ValidationException) when calling the UpdateItem operation: ExpressionAttributeValues contains invalid value: One or more parameter values were invalid: An string set  may not be empty for key <SOME KEY>
```

#### `dynamo_dump -> None`

At first I considered letting the `dynamo_dump` function return a `None` instead of an empty set:

```
class Set(...):
    ...
    def dynamo_dump(self, value, *, context=None, **kwargs):
        dump = self.typedef.dynamo_dump
        return [dump(v, context=context, **kwargs) for v in sorted(value)] or None
```

which serializes to `{"SS": None}` and results in a slightly different error message:

```
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid type for parameter ExpressionAttributeValues.<SOME KEY>.SS, value: None, type: <class 'NoneType'>, valid types: <class 'list'>, <class 'tuple'>
```

#### `ConditionRenderer.update`

So the problem is actually that we can't have this be part of the `SET` section of the update; we need it to be in `REMOVE`.  Over to `condition.ConditionRenderer.update`.

Every tuple of `(column, value)` comes in from `tracking.get_update` that comes in as part of `SET` gets dumped through `ConditionRenderer.refs` -> `ConditionRenderer.value_ref` which will get our `{"SS": <something>}` from the Type.  Assuming we broke that pattern to signal "this value is empty!" we still need to get this value out of `SET` and into `REMOVE`.

#### `tracking.get_update`

Almost there.  Here's the relevant bits of `get_update`:

```python
for column in _obj_tracking[obj]["marked"]:
    if column in key:
        continue
    value = getattr(obj, column.model_name, None)  # <- OH NOOOOO!
    if value is not None:
        diff["SET"].append((column, value))
    # None (or missing, an implicit None) expects the
    # value to be empty (missing) in Dynamo.
    else:
        diff["REMOVE"].append(column)
```

That `getattr` is returning the loaded python version of the object; in our sample above, that's `set()`.  Which means it doesn't matter if we signal that the dumped form is `None`, because it's already pushed into the wrong segment of the update.

#### Clean up tracking

There are a few parts.  First, drop `get_update` from tracking and let `ConditionRenderer` do the work.  Instead, tracking should expose `marked_columns(obj) -> List[column.Column]` and let the renderer decide what needs to happen.

#### Drop `None` protections from `types.Type`

Going to need some docs work on this one, too.

In short, it's not sufficient to let `Type._dump` return `None` and call it a day, because nested types will end up having a bunch of Nones.  Consider the following document:

```python
typedef = types.TypedMap(Set(Integer))

context = {"engine": engine}
value = {"foo": set(), "bar": set()}
```

Assuming `Type._dump` can return a raw `None`, when the `TypedMap` recursively dumps its `.values()` it will end up with:

```python
{
    "M": {
        "foo": None,
        "bar": None
    }
}
```

Those `None`s are still incorrectly formatted parameters; they need to be `{DYNAMO_TYPE: VALUE}` and we already know that `{"SS": None}` won't work either.

So, recursive types will need to filter out `None`s and potentially return `None` when al of their inner values are `None`.
